File: App/Resources/AlgorithmDetails/binomialsmoothsort/binomialsmoothsort.cpp
#include <cstdio>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  printf("[");
  if (size > 0) {
    printf("%d", items[0]);
    for (int i = 1; i < size; i++) {
      printf(", %d", items[i]);
    }
  }
  printf("]");
}

int height(int node);
void thrift(int node, bool parentFlag, bool rootFlag);

void sort(int arr[], int n) {
  int node = 1;
  while (node < n) {
    thrift(node, node % 2 == 1, (node + (1 << height(node))) >= n);
    node++;
  }

  node -= (node - 1) % 2;
  while (node > 2) {
    for (int child = height(node) - 1; child >= 0; child--) {
      thrift(node - (1 << child), false, true);
    }
    node -= 2;
  }
}

int height(int node) {
  int count = 0;
  while ((node >> count) % 2 == 1) {
    count++;
  }
  return count;
}

void thrift(int node, bool parentFlag, bool rootFlag) {
  bool isRoot = rootFlag && (node >= (1 << height(node)));
  if (!isRoot && !parentFlag) {
    return;
  }

  int choice = height(node) - (isRoot ? 0 : 1);
  if (parentFlag) {
    int bound = choice;
    for (int child = bound - 1; child >= 0; child--) {
      if (array[node - (1 << choice)] <= array[node - (1 << child)]) {
        choice = child;
      }
    }
  }

  if (array[node - (1 << choice)] <= array[node]) {
    return;
  }

  std::swap(array[node], array[node - (1 << choice)]);
  int nextNode = node - (1 << choice);
  thrift(nextNode, nextNode % 2 == 1, choice == height(node));
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: Modules/BuiltInAlgorithms/Sources/BinomialSmoothSort.swift
import AlgorithmKit
import SortEngineKit

public struct BinomialSmoothSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "binomialsmoothsort")
  public let metadata = AlgorithmMetadata(
    displayName: "Binomial Smooth Sort",
    category: .selection,
    sizeRange: 16...256,
    stable: false,
    timeComplexity: ComplexityBounds(
      best: "O(n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(1)",
    iconName: "wind"
  )
  public init() {}

  /// A Smoothsort-family algorithm built on the same implicit-binomial-heap-in-an-array idea as
  /// `BinomialHeapSort`, but 0-indexed and recursive (`thrift`). `height(node)` counts `node`'s
  /// trailing set bits, telling `thrift` which sibling subtree to look at. `thrift(node, parent,
  /// root)` decides whether an earlier subtree root beats `node` and, if so, swaps it up and
  /// recurses into the vacated spot.
  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    func height(_ node: Int) -> Int {
      var count = 0
      while (node >> count) % 2 == 1 {
        count += 1
      }
      return count
    }

    func thrift(_ node: Int, _ parentFlag: Bool, _ rootFlag: Bool) {
      let isRoot = rootFlag && (node >= (1 << height(node)))
      guard isRoot || parentFlag else { return }

      var choice = height(node) - (isRoot ? 0 : 1)
      if parentFlag {
        for child in stride(from: choice - 1, through: 0, by: -1)
        where !engine.compare(node - (1 << choice), node - (1 << child), by: (>)) {
          choice = child
        }
      }
      guard engine.compare(node - (1 << choice), node, by: (>)) else { return }

      engine.swap(node, node - (1 << choice))
      let nextNode = node - (1 << choice)
      thrift(nextNode, nextNode % 2 == 1, choice == height(node))
    }

    var node = 1
    while node < n {
      thrift(node, node % 2 == 1, (node + (1 << height(node))) >= n)
      node += 1
    }

    node -= (node - 1) % 2
    while node > 2 {
      for child in stride(from: height(node) - 1, through: 0, by: -1) {
        thrift(node - (1 << child), false, true)
      }
      node -= 2
    }
  }
}
